Introduce a way of running external programs

shaded-enmity opened the pull request:
This function provides stable interface across Python2 and Python3.
Besides, the main difference between this and the various `subprocess`
related functions is that we can easily "tee" the output to stdout/stderr
while buffering it internally for auditing.

Note that the `_call` function should provide internal interface for Leapp and that we should provide additional "facade" on top of that function to simplify the interface (and perhaps mimic the `subprocess` interface).

You can easily test this function by calling the file directly:

```
$ python2 call.py bash -c 'echo 123 ; (>&2 echo err) ; exit 255'
123
err
{'exit_code': 255,
 'pid': 16412,
 'signal': 0,
 'stderr': u'err\n',
 'stdout': u'123\n'}
$ python3 call.py dnf update
Error: This command has to be run under the root user.
{'exit_code': 1,
 'pid': 13591,
 'signal': 0,
 'stderr': u'Error: This command has to be run under the root user.\n',
 'stdout': u''}
```

pirat89 commented on leapp/libraries/common/call.py:
```diff
@@ -0,0 +1,119 @@
+import sys
+
+
+STDOUT = 1
```
Wouldn't be better to use: `sys.stderr.fileno()` and `sys.stdout.fileno()`

pirat89 commented:
@vinzenz I guess this one should be under `stdlib`, right? `common` is meant to be part of *leapp-repository*.

vinzenz commented on leapp/libraries/common/call.py:
```diff
@@ -35,7 +39,7 @@ def multiplex_reads(ep, fds, callback_raw, callback_linebuffered, timeout=1, buf
             buf[fd] += read
 
-    # process leftoevers from line buffering
+    # Process leftoevers from line buffering
```
leftoevers => leftovers

Rezney commented:
+1 for adding the comments

vinzenz commented on leapp/libraries/common/call.py:
```diff
@@ -0,0 +1,135 @@
+
+        read = multiplex_reads(
+            ep,
+            [stdout, stderr],
```
I guess that it won't matter that much, but you never change this list, I think it's a good idea to have it a tuple, lists are just noice IMHO but well I will let this be up to you.

shaded-enmity commented on leapp/libraries/common/call.py:
```diff
@@ -0,0 +1,135 @@
+    return buf
+
+
+def _call(command, callback_raw=lambda fd, value: None, callback_linebuffered=lambda fd, value: None,
```
All should be private FTTB

shaded-enmity commented on leapp/libraries/common/call.py:
```diff
@@ -0,0 +1,135 @@
+    return data.decode('utf-8')
+
+
+def multiplex_reads(ep, fds, callback_raw, callback_linebuffered, timeout=1, buffer_size=72):
```
Changed size to 80 and added notice into docstring

shaded-enmity commented:
@vinzenz @Rezney @pirat89 Hopefully addressed all the comments. (besides the testing which i'm going to take care of now)

I also extended the code so that it is possible to pass `stdin` to the child process as well (either directly as a string, or a file descriptor).

shaded-enmity commented:
Ok folks, I think that we're good to go here.

As for the final facade I'd start with the following (the implementation part is pseudocode):
```python
def call(command, stdout=True, stderr=True, stdin=fd/str, split=True):
     Audit.add('COMMAND_START', command, bytes(stdin), user.uid, user.gid)
     def linebuffer_callback(fd, data):
         log(data)
     r = _call(command, linebuffer_callback=linebuffer_callback, stdin=stdin)
     Audit.add('COMMAND_FINISH', r['stdout'], r['stderr'], r['error_code'], r['pid'], r['signal'])
     if not stdout:
         del r['stdout']
     elif split:
         r['stdout'] = r['stdout'].splitlines()
     if not stderr:
         del r['stderr']
     elif split:
         r['stderr'] = r['stderr'].splitlines()
         
     return r
```
If either stdout/stderr are False, their output is not included in the return dictionary (just the error_code, pid and signal if both are off)

This should be compatible with the old interface, while providing new means of control for users.
